Feature tests as GitHub actions

felixsch opened the pull request:
## Description

Bring feature tests into Github and run them as Github actions, rather then relying on our internal CI pipelines with its complicated setup.

This change will:
- Run feature tests in github actions
- Auto build the ci-container on master push
- Add scriptlets to easily build rpms and run feature tests

### Run the pipeline locally

- requires you to run a mariadb on your host machine
- Checkout of the branch into <RMT>

```
$ cd <RMT>/ci/
$ docker build --no-cache -t rmt-devel -f Dockerfile .
$ docker run --rm -it -e SCC_USERNAME="<username>" -e SCC_PASSWORD="<password>" --network=host -v $(pwd):/usr/src/rmt-server rmt-devel

> ./ci/rmt-build-rpm 
# expect: in /usr/src/rmt-server/tmp/artifacts/ are built rpms

> ./ci/rmt-configure
# expect: installs and configures rmt from the rpm

> ./ci/rmt-run-feature-tests
# expect: runs the feature tests
```

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356839/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 9a56c1e5add72b85e75b4d8ab5242e4429dde7fe into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356840/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged bc44b2a7476fd65f4ce5ffa7efd3e75a28998ce2 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356863/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 99449b8b9d921be3498a4b6b30d5535f3a9a6f86 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

josegomezr commented on ci/rmt-configure:
```diff
@@ -0,0 +1,78 @@
+ARTIFACT_DIR=${ARTIFACT_DIR:-$SOURCE/tmp/artifacts}
+
+# Current version we are working with
+VERSION=$(ruby -e "require '$SOURCE/lib/rmt.rb'; print RMT::VERSION")
```
```suggestion
VERSION=$(ruby -r "$SOURCE/lib/rmt" -e 'print RMT::VERSION')
```

josegomezr reviewed (commented):
Have a look at the comments.

and:

- would it make sense to have central bash file with the functions and common prelude (`SOURCE=...`) and then just `source my-prelude.sh`?
- Given that there are more bash scripts, would it make sense to pull `shelllint` into the actions for some sanity checks?

felixsch commented on .github/workflows/features.yml:
```diff
@@ -0,0 +1,75 @@
+
+      - name: move source to /usr/src/rmt-server
+        run: |
+          [ -d $SOURCE ] && rm -r $SOURCE
```
it creates the directory to let it mount later. I'm not sure if this is actually required.

felixsch replied:
I let this stay for now but will reiterate when the whole ci container pipeline is up and running!

felixsch commented on .github/workflows/features.yml:
```diff
@@ -0,0 +1,75 @@
+
+env:
+  SOURCE: /usr/src/rmt-server
+  SCC_USERNAME: ${{ secrets.SCC_USERNAME }}
+  SCC_PASSWORD: ${{ secrets.SCC_PASSWORD }}
```
I checked. Secrets are not exposed when running actions from forks.
vvv from the secrets page.
![image](https://github.com/SUSE/rmt/assets/4460715/8b9ee7d0-7ec4-41ba-9c5b-94412d282344)

I will look into narrowing it down further

felixsch replied:
I added a comment paragraph explaining the current situation. Do you think we need more hardening? @josegomezr

josegomezr replied:
I don't know for sure. But we can definitely add it as we discover it

felixsch replied:
of course. I will let the current configuration as it is and we adapt accordingly

felixsch commented on Makefile:
```diff
@@ -10,7 +10,7 @@ clean:
 
 man:
-	bundle exec ronn --roff --pipe --manual RMT MANUAL.md > rmt-cli.8 && gzip -f rmt-cli.8
+	ronn --roff --pipe --manual RMT MANUAL.md > rmt-cli.8 && gzip -f rmt-cli.8
```
this is installed via rpm package. Idea was/is to not have to run `bundle install` when `make dist` Do you think this is a problem?

josegomezr replied:
ah, so it comes baked in the base container? In that case: even better, we can resolve this

felixsch commented on .github/workflows/features.yml:
```diff
@@ -0,0 +1,75 @@
+        with:
+          fetch-depth: 0
+
+      - name: move source to /usr/src/rmt-server
+        run: |
+          [ -d $SOURCE ] && rm -r $SOURCE
+          cp -r $GITHUB_WORKSPACE $SOURCE
```
Don't get me started on this stuff. There are multiple bug reports open against `checkout` which does not allow to alter the checkout path outside of original $GITHUB_WORKSPACE. They even suggest to move the result until the fix is provided.

josegomezr replied:
I imagine :joy: I'm marking this as resolved, the current state works

josegomezr commented on .github/workflows/features.yml:
```diff
@@ -0,0 +1,75 @@
+          --health-timeout=5s
+          --health-retries=3
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v4
```

we're in v4 rn

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356886/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 2ba89d1f4fc2722c44ac9e428414514cccfe7163 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356888/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment. (Merged 212be2268402ab93861572044d87a1d61185436c into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

suse-tests-fail commented:
Prophet reports failure.
https://ci.suse.de/job/scc-RMT-integration-tests/356897/console
If the given link has expired,you can force a Prophet rerun by just deleting this comment.
Failed fast in project rmt on step docker-compose-build-images. (Merged 196e4f457c0de5f3031358615ec6d50c94b78db1 into ecf5efb3d9dbfce20f8d302a5ae4db03e08f611d)

felixsch commented:
I had a discussion about the *prelude* for the scripts. I decided against it because I favor self-contained scripts over configuration repition. @josegomezr offered to make a follow up PR if he deems it necessary to make the prelude happen.

I added the shell linting and fixed bash scripts in `ci/*`

ngetahun reviewed (approved):
LGTM!

josegomezr reviewed (approved):
LGTM!
